### PR TITLE
Add gitoid crate

### PR DESCRIPTION
**Pull request by alilleybrinker**
This is an initial effort at splitting the `gitbom` crate into two crates: `gitbom`, which provides the overall BOM structure, and `gitoid`, which only handles the construction of gitoids. This PR also includes some changes to the `GitOid` API as it existed in the singular crate to make the API more consistent between sync and async, including removal of the `Source` type and delegation of any mass construction of `GitOid`s to users of the API, rather than baking the iteration into the API itself.

This also includes some cleanup / consistency improvements to the test code, as well as changing the methods and constants for `HashAlgorithm` and `NUM_HASH_BYTES` to be `pub(crate)` instead of `pub`.

The individual commit messages try to explain more concretely.

### CODE REVIEW

**Comment by alilleybrinker** [comment]
Okay, rebased with signoff and force-pushed, and it looks like the DCO check is happening now!

**Comment by alilleybrinker** [comment]
For whomever reviews this: it's probably easier to look at the individual commits and their commit messages rather than looking at the overall view of the changes.

**Comment by alilleybrinker** [comment]
Okay, broken test should be fixed, and lack of signoff on the new commit should be fixed as well.

**Comment by alilleybrinker** [comment]
I should also note that someone ought to publish the new crate if/once this is accepted. Also a question about changing the import in `gitbom` after that.

File: gitoid/src/bin/gitoid.rs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+fn main() {
```
Comment (nellshamrell) [inline gitoid/src/bin/gitoid.rs:1]: Did you intend to leave this file in?

Comment (alilleybrinker) [reply gitoid/src/bin/gitoid.rs:1]: Ah, it can be removed. Mostly it was a placeholder so I don't forget to go back and implement a CLI.

Comment (nellshamrell) [reply gitoid/src/bin/gitoid.rs:1]: Not sure if this is what you are thinking of, but I've been playing with a CLI for this here https://github.com/nellshamrell/gitbom-rs-cli. I created this mainly to exercise this library as we developed it and am certainly not attached to it being where it is.

(We had discussed whether it made sense to include the cli in this crate or to keep it separate - I'm definitely open to considering including it within this crate if that would make more sense)

Comment (alilleybrinker) [reply gitoid/src/bin/gitoid.rs:1]: Alright, cool. I'll take a look, and we can have a separate discussion later. For now, I'll remove the empty binary crate.

**Comment by nellshamrell** [comment]
Just did a thorough review of this - this PR is outstanding! Once the file at gitoid/src/bin/gitoid.rs is removed, this will be good to merge!

**Comment by alilleybrinker** [comment]
Alright, first pass forgot to include the sign-off, but that's fixed. Should be good to merge now! Thanks Nell!

**Review by fkautz — APPROVED** [review]

**Pull request merged**